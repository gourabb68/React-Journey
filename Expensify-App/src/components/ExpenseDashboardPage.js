
import React from 'react';
import {Link, NavLink} from 'react-router-dom';
const ExpenseDashboradPage = () =>{
    return (
        <div>
        this is from my dashboard page
        <div>
        <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard </NavLink>           
        <NavLink to="/create"activeClassName="is-active">Create Page</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help Page</NavLink>
        </div>
         
    </div>
    );
 
};
export default ExpenseDashboradPage;