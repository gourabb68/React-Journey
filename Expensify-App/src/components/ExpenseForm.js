import React from 'react';
////moment for working with  date 
import moment from 'moment';
//import date picker library
import {SingleDatePicker} from 'react-dates';
//import cssfile for date picker library
// import 'react-dates/lib/css/_datepicker.css';
const now = moment();
// console.log(now)//get moment object
//get date
// console.log(now.format('MMM Do, YYYY'))//Jun 8th, 2020

class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state ={
            description: props.expenses? props.expenses.description: '',
            note: props.expenses? props.expenses.note:'',
            amount: props.expenses?(props.expenses.amount/100).toString():'',
            createdAt: props.expenses?moment(props.expenses.createdAt):moment(),
            calenderFocused: false,
            error: ''
        }
    }

onDescriptionChange =(e) =>{
    const description = e.target.value;
    this.setState(()=>({description}))
}
onNoteChange =(e)=>{
    const note = e.target.value;
    this.setState(()=>({note}))

}
onAmountChange =(e)=>{
    const amount = e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){ //copy paste reg expression between // 
        //!amount if for if user can delete the amount
        //creating regular expression to allow user take 
        //two value after decimal like 55.55 not 55.5555
    this.setState(()=>({amount}))
    }
}
///https://github.com/airbnb/react-dates go to single date picker
onDateChange =(createdAt) =>{
    if(createdAt){
   this.setState(()=> ({createdAt}))
    }
};

onFocusChange = ({focused})=>{
   this.setState(()=>({calenderFocused : focused}))
};
onSubmit = (e)=>{
    //prevent from full page refresh
    e.preventDefault();
    if(!this.state.description || !this.state.amount){
        this.setState(()=>({error: 'Please provide Description and amount'}))
        //set error state equal to 'Please provide dec and amount
        // console.log(this.state.error)
    }else{
        this.setState(()=>({error: ''}))
        this.props.onSubmit({
            description: this.state.description,
            amount: parseFloat(this.state.amount,10)*100, //convert string to no
            createdAt: this.state.createdAt.valueOf() ,//moment to unix timestamp
            note: this.state.note
        })
    }
}
    render(){
        return (
                <form className="form" onSubmit={this.onSubmit}>
                 {this.state.error && <p className="form__error">{this.state.error}</p>}
                    <input className="text-input" type='text'
                    placeholder='Description'
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}/>

                    <input className="text-input" type='number'
                    placeholder='Amount'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    />

                    <textarea className="textarea" placeholder=
                    'Add a note for your expense'
                    value={this.state.note}
                    onChange={this.onNoteChange}>
                        
                 

                    </textarea>
                   
                    {/* {console.log(this.state.amount)} */}

                    <SingleDatePicker
                    date ={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused = {this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1} //showing no of month in calender
                    isOutsideRange= {() => false} //it will make every single day available for us like past
                    />
                    <div >
                     <button className="button">Save Expense</button>
                     </div>
                </form>
        );
    }
}
export default ExpenseForm;