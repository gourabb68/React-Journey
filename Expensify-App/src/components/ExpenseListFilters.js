import React from 'react';
import { connect } from 'react-redux';
//we need setTextFilter action to dispatch to redux store
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null,

    }
    onDatesChange = ({ startDate, endDate }) => {
        if (startDate) {
            console.log('start' + startDate);
            this.props.dispatch(setStartDate(startDate));
        }
        if (endDate) {
            console.log('end' + endDate);
            this.props.dispatch(setEndDate(endDate));
        }


    };
    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }));
    }
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input type='text' className="text-input" placeholder="Search Expense" value={this.props.filters.text} onChange={(e) => {
                            //we have dispatch option available as we connect to redux store
                            this.props.dispatch(setTextFilter(e.target.value));//passing value given by user in text file                  
                        }} />
                    </div>
                    <div className="input-group__item">
                        <select className="select"onChange={(e) => {
                            console.log(e.target.value)
                            if (e.target.value === 'date') {
                                props.dispatch(sortByAmount());
                            } else if (e.target.value === 'amount') {
                                this.props.dispatch(sortByDate());
                            }
                        }}>
                            <option  value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calenderFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}


                        />
                    </div>
                </div>





            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}
export default connect(mapStateToProps)(ExpenseListFilters);