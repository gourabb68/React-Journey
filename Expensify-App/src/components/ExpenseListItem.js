import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
//first value is dispatch
export const ExpenseListItem = ({ description,amount,createdAt,id})=>{
    return (
      <div>
        <Link to={`/edit/${id}`} activeClassName="is-active" exact={true}>
          <h3>{description}</h3>
          </Link>
          <p>
            {numeral(amount/100).format('$0,0.00')} - 
            {moment(createdAt).format('MMMM Do, YYYY')}
          </p>
          {/* //creating button to remove expense} */}
         
          
      </div>
    );
}


export default connect()(ExpenseListItem);