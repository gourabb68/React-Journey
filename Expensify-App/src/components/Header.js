import React from 'react';
import {Link, NavLink} from 'react-router-dom';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth';
export const Header =(props) =>{
    return (
     <header>
     <h1>Expensify</h1>
     <div>
     <NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard </NavLink>           
        <NavLink to="/create"activeClassName="is-active">Create Page</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help Page</NavLink>
        <button onClick={props.startLogout}>Logout</button>
     </div>
     </header>
    ); 
 };

 const mapDispatchToProps =(dispatch)=>({
    startLogout : () => dispatch(startLogout())
 })
 export default connect(undefined,mapDispatchToProps)(Header);