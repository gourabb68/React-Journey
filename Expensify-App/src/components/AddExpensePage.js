import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import { startAddExpense} from '../actions/expenses';


export class AddExpensePage extends React.Component {
    onSubmit=(expenses)=>{
        // props.dispatch(addExpense(expenses));
        this.props.startAddExpense(expenses);
        this.props.history.push('/');//switch me to the dashboard page
    }
    render (){
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                         <h2 className="page-header__title1">Add Expense</h2>
                         </div>
                         </div>
                   <div className="content-container">
                     <ExpenseForm 
                      onSubmit={this.onSubmit}
                    // onSubmit={(expenses)=>{
                    //     // props.dispatch(addExpense(expenses));
                    //     this.props.onSubmit(expenses);
                    //     this.props.history.push('/');//switch me to the dashboard page
                    // }}
                    />
                    </div>
                     
                </div>
        )
    }
}
// const AddExpensePage = (props) =>{
//     console.log(props);
//     return (
//         <div>
//             <h1>Add Expense</h1>
//         <ExpenseForm 
//         onSubmit={(expenses)=>{
//             // props.dispatch(addExpense(expenses));
//             props.onSubmit(expenses);
//             props.history.push('/');//switch me to the dashboard page
//         }}
//         />
         
//     </div>
//     );
 
// };


const mapDispatchToProps =(dispatch) =>({
    startAddExpense: (expenses)=> dispatch(startAddExpense(expenses))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage);

