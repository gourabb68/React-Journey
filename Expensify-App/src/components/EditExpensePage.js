import React from 'react';
import {Link, NavLink} from 'react-router-dom';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm'
import {startEditExpense} from '../actions/expenses';
import {startRemoveExpense} from '../actions/expenses';


export class EditExpensePage extends React.Component{
    onSubmit = (expenses)=>{                   
            this.props.startEditExpense(this.props.expenses.id,expenses);
           this.props.history.push('/dashboard');//switch me to the dashboard page      
    };

    onRemove = ()=>{
               
        this.props.startRemoveExpense({id:this.props.expenses.id});
        this.props.history.push('/dashboard');
    }

    render(){
        return(
            <div>
            <div className="page-header">
            <div className="content-container">
               <h2 className="page-header__title1">Edit Expense</h2>
            </div>
            </div>
            <div className="content-container">
            <ExpenseForm 
            expenses={this.props.expenses}
            onSubmit={this.onSubmit}
            />
              <button className="button button__secondary" onClick={this.onRemove}>Remove Expense</button>
              </div>
        </div>
        )
    }
}

const mapDispatchToProps =(dispatch,props) =>({
   
    startEditExpense: (id,expenses)=>dispatch(startEditExpense(id,expenses)),
    startRemoveExpense: (id)=> dispatch(startRemoveExpense({id}))
    
})

const mapStoreToProps =(state,props)=>{
return {
    expenses: state.expenses.find((expense)=>{
        return expense.id === props.match.params.id;
    })
};
}
export default connect(mapStoreToProps,mapDispatchToProps)(EditExpensePage);
