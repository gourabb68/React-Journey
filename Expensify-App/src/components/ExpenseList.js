import React from 'react';
//use to connect redux store with component
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'
export const  ExpenseList = (props) =>{
    return (
       
        <div className="content-container">
        <div className="list-header">
          <div className="show-for-mobile">Expenses</div>
          <div className="show-for-desktop">Expense</div>
          <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
          {
            props.expenses.length === 0 ? (
              <div className="list-item list-item__message">
                <span>No expenses</span>
              </div>
            ) : (
                props.expenses.map((expense) => {
                  return <ExpenseListItem key={expense.id} {...expense} />;
                })
              )
          }
        </div>
      </div>
    );
   
}
//creating higher order component using connect
//like we have our own function and passing regular component
// const ConnectedExpensesList connect(here we have to pass function that 
// return object which is subset of store means what info we need to c
// onnect from store)(pass regular component to make higer order component)
//one way
// const ConnectedExpensesList= connect((state)=>{//this return object can be accessed through props of regular comp
//     return {
//         expenses: state.expenses
//     };
// })(ExpenseList)
//advance way
//creating a funciton explicitly to get data from store
const mapStateToComponent = (state)=>{//this return object can be accessed through props of regular comp
    return {
        expenses: selectExpenses(state.expenses,state.filters)
        
    };
}
const ConnectedExpensesList= connect(mapStateToComponent)(ExpenseList)




export default ConnectedExpensesList; 