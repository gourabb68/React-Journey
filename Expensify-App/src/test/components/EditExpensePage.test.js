import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
const expenses =  [{
    id: 'abcdef',
    description: 'januray rent',
    note: 'This was the final payment',
    amount: 54200,
    createdAt: 0
}]


let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expenses={expenses[0]}
    />
  );
});

test('should render EditExpensePage', () => {
  expect(wrapper).toMatchSnapshot();
});

// test('should handle editExpense', () => {
//   wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
//   expect(history.push).toHaveBeenLastCalledWith('/');
//   expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
// });

// test('should handle removeExpense', () => {
//   wrapper.find('button').simulate('click');
//   expect(history.push).toHaveBeenLastCalledWith('/');
//   expect(removeExpense).toHaveBeenLastCalledWith({
//     id: expenses[0].id
//   });
// });
