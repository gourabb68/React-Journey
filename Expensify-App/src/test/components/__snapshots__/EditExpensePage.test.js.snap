// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title1"
      >
        Edit Expense
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expenses={
        Object {
          "amount": 54200,
          "createdAt": 0,
          "description": "januray rent",
          "id": "abcdef",
          "note": "This was the final payment",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button__secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
